### ISSUE TITLE
IDE0066 incorrectly handles nullable value types

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#37950]
**Version Used**: 
3.3.0-beta3-19406-05+a1905991543bed104f7f7f0842aca2b65d263b87

**Steps to Reproduce**:
```C#
class Program
{
    public static void Main() { }

    public static bool? GetBool(string name)
    {
        switch (name)
        {
            case "a": return true;
            case "b": return false;
            default: return null;
        }
    }
}
```

**Expected Behavior**:
IDE0066 refactors this into something that compiles, e.g.
```C#
class Program
{
    public static void Main() { }

    public static bool? GetBool(string name)
    {
        return name switch
        {
            "a" => true,
            "b" => false,
            _ => (bool?)null,
        };
    }
}
```

**Actual Behavior**:
IDE0066 refactors this into the following code that doesn't compile:
```C#
class Program
{
    public static void Main() { }

    public static bool? GetBool(string name)
    {
        return name switch
        {
            "a" => true,
            "b" => false,
            _ => null,
        };
    }
}
```
This fails with the error "error CS0037: Cannot convert null to 'bool' because it is a non-nullable value type".

**Comment by alrz** [comment]
Fixed in #37052

**Comment by alrz** [comment]
note that this would be still an issue if there is no target-type until https://github.com/dotnet/csharplang/issues/33 is implemented, but we wont suggest 0066 in those cases.

**Comment by stephentoub** [comment]
> Fixed in #37052

VS says my tools are from a1905991543bed104f7f7f0842aca2b65d263b87, and according to git log, that includes fbdb480b187060347aa63050e6fc288d50eeda34.  Am I misinterpreting?

**Comment by alrz** [comment]
perhaps I am. According to the feature spec https://github.com/dotnet/csharplang/issues/2389 I expect this to work - there's no natural type but all arms are implicitly convertible to the target type.

cc @gafter

**Comment by alrz** [comment]
The fix for IDE is in review #38007, in case this is the intended behavior.

**Comment by gafter** [comment]
Th problem is that there *is* a common type according to the current spec/implementation - because only two switch arms have expressions with types, and those are the same type.  The spec and implementation should be amended to require that all arms are convertible to that common type in order for it to be considered the natural type of the switch.

**Comment by alrz** [comment]
>The spec and implementation should be amended to require that all arms are convertible to that common type in order for it to be considered the natural type of the switch.

I'll note that https://github.com/dotnet/csharplang/issues/33 will address this issue by taking `null` into account in the best-type inference, thought it is probably an uninitave behavior at the moment.

**Comment by gafter** [comment]
Yes, and it will take lambdas, method groups, and other typeless things into account too.

**Comment by alrz** [comment]
So if I understand correctly, "nullable-enhanced common type" would not help with the following 

```cs
(int?, int) t = b switch { true => (null, 1), false => (1, 2) };
```
but it should actually work because there's no natural type for the switch expression.

**Comment by gafter** [comment]
I think that's right.  However, it should work whether or not there is a natural type for the switch expression.

**Comment by gafter** [comment]
I am going to aim to fix the compiler behavior in 16.4.

**Comment by gafter** [comment]
Compiler part moved to https://github.com/dotnet/roslyn/issues/38226
